ygot/path_types.go

```diff
@@ -47,9 +54,46 @@ type NodePath struct {
 	p             PathStruct
 }
 
-// ResolvePath is a helper which returns the root PathStruct and absolute path
-// of a PathStruct node.
-func ResolvePath(n PathStruct) (PathStruct, []*gpb.PathElem, []error) {
+// fakeRootPathStruct is an interface that is implemented by the fake root path
+// struct type.
+type fakeRootPathStruct interface {
+	PathStruct
+	Id() string
+	CustomData() map[string]interface{}
+}
+
+func NewDeviceRootBase(id string) *DeviceRootBase {
+	return &DeviceRootBase{NodePath: &NodePath{}, id: id, customData: map[string]interface{}{}}
+}
+
+// DeviceRootBase represents the fakeroot for all YANG schema elements.
+type DeviceRootBase struct {
+	*NodePath
+	id string
+	// customData is meant to store root-specific information that may be
+	// useful to know when processing the resolved path. It is meant to be
+	// accessible through a user-defined accessor.
+	customData map[string]interface{}
+}
+
+// Id returns the device ID of the DeviceRootBase struct.
+func (d *DeviceRootBase) Id() string {
+	return d.id
+}
+
+// CustomData returns the customData field of the DeviceRootBase struct.
+func (d *DeviceRootBase) CustomData() map[string]interface{} {
+	return d.customData
+}
+
+// SetCustomDataKey modifies an entry in the customData field of the DeviceRootBase struct.
+func (d *DeviceRootBase) SetCustomDataKey(key string, val interface{}) {
```

Comment:
Maybe it's not standard in ygot, but I'm used to seeing the verb "put" for maps. And then maybe you no longer need the "key" part, i.e. just "PutCustomData"